src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -0,0 +1,165 @@
+-- Complexity: /O(log(n))/ time.
+--
+-- @
+-- hasEdge x y ('fromGraph' 'Algebra.Graph.empty')                                 == False
+-- hasEdge x y ('fromGraph' ('Algebra.Graph.edge' (Left x) (Right y)))             == True
+-- hasEdge 1 2 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Left 2)))              == False
+-- hasEdge 2 3 ('fromGraph' ('Algebra.Graph.edge' (Left 1) (Right 2)))             == False
+-- hasEdge x y ('fromGraph' ('Algebra.Graph.Overlay' g ('Algebra.Graph.edge' (Left x) (Right y)))) == True
+-- @
+hasEdge :: (Ord a, Ord b) => a -> b -> AdjacencyMap a b -> Bool
+hasEdge u v (BAM m _) = ((Set.member v) <$> (u `Map.lookup` m)) == Just True
+
+-- | Adds all edges needed to make the graph undirected.
+-- Complexity: /O(m log n)/
+--
+-- /Note: this function is for internal use only./
+addReverseEdges :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AM.AdjacencyMap (Either a b)
+addReverseEdges m = AM.overlay m $ AM.edges [ (v, u) | (u, v) <- AM.edgeList m ]
+
+-- | Constructs a bipartite 'Bipartite.AdjacencyMap' from
+-- 'Algebra.Graph.AdjacencyMap' with given part identifiers, adding all needed
+-- edges to make the graph undirected and removing all edges inside one part.
+-- Complexity: /O(m log n)/
+--
+-- @
+-- 'leftToRight' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.empty'))                                        == Map.'Map.empty'
+-- 'leftToRight' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Right 2)))                      == Map.'Map.singleton' 1 (Set.'Set.singleton' 2)
+-- 'leftToRight' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.edge' (Left 1) (Left 2)))                       == Map.'Map.empty'
+-- 'rightToLeft' (fromAdjacencyMap ('Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 1), (Left 1, Right 2)])) == Map.'Map.fromAscList' [(1, Set.'Set.singleton' 1), (2, Set.'Set.singleton' 1)]
+-- @
+fromAdjacencyMap :: (Ord a, Ord b) => AM.AdjacencyMap (Either a b) -> AdjacencyMap a b
+fromAdjacencyMap m = BAM (Map.fromAscList [ (u, setRights vs) | (Left  u, vs) <- Map.toAscList (AM.adjacencyMap $ addReverseEdges m)])
+                         (Map.fromAscList [ (u, setLefts  vs) | (Right u, vs) <- Map.toAscList (AM.adjacencyMap $ addReverseEdges m)])
+    where
+        setRights = Set.fromAscList . rights . Set.toAscList
+        setLefts  = Set.fromAscList . lefts  . Set.toAscList
+
+-- | Constructs an 'Algrebra.Graph.AdjacencyMap' from a bipartite
+-- 'AdjacencyMap'.
+-- Complexity: /O(m log n)/
+--
+-- @
+-- toAdjacencyMap (fromGraph 'Algebra.Graph.empty')                     == 'Algebra.Graph.AdjacencyMap.empty'
+-- toAdjacencyMap (fromGraph ('Algebra.Graph.vertex' $ Left 1))         == 'Algebra.Graph.AdjacencyMap.vertex' 1
+-- toAdjacencyMap (fromGraph ('Algebra.Graph.edge' (Left 1) (Right 2))) == 'Algebra.Graph.AdjacencyMap.edges' [(Left 1, Right 2), (Right 2, Left 1)]
+-- @
+toAdjacencyMap :: (Ord a, Ord b) => AdjacencyMap a b -> AM.AdjacencyMap (Either a b)
```

Comment:
After checking other modules, I think this should be named `fromBipartite` for consistency.